File: OZprivate/ServerScripts/Utilities/getBestEOLdoID.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
''' Given a csv input file containing EOL page numbers, return a file of the best pd|cc-by|cc-by-sa picture ids and the common names. It is useful to include OTT numbers for each EOL id, so there can be multiple columns in each line of the file, one of which is an EOL id. The column number (1,2,3 etc) containing the EOL id can be specified using -c X

e.g. list of all leaves using 
 cut -f 2,4 -d, data/output_files/ordered_leaves.csv | ServerScripts/Utilities/getBestEOLdoID.py -c 2 > output.txt

 The simplest use of this script is e.g. 
      ./getBestEOLdoID.pl pageIDs.txt output.txt
'''

import os
import sys
import re
import csv
import argparse
import requests
from requests.packages.urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter
import html

def warn(*objs):
  print(*objs, file=sys.stderr)

def getPageInfo(input_ids, sess):
    url = "http://eol.org/api/pages/1.0.json"; #see http://eol.org/api/docs/pages
    APIkey="0e8786f5d94e9587e31ed0f7703c9a81f3036c7f"; #replace with your own API key.
    pages_params = {
        'batch'          : 'true',
        'key'            : APIkey,
        'images_per_page': 1,     # only look at the first images
        'videos_per_page': 0,
        'sounds_per_page': 0,
        'maps_per_page'  : 0,
        'texts_per_page' : 0,
        'iucn'           : 'false',
        'licenses'       :'pd|cc-by|cc-by-sa', #change this to get objects distributed under different licences
        'details'        : 'false',
        'references'     : 'false',
        'synonyms'       : 'false',
        'taxonomy'       : 'false',
        'language'       : args.language,
        'cache_ttl'      : 2500, #only cache for X secs
    }
    input_ids = [id for id in input_ids] #convert to array
    eol_ids = [str(id) for id in input_ids]
    if args.verbose:
        warn('checking ' , eol_ids)

    try:
        #If 'vetted' is given a value of '1', then only trusted content will be returned. If 'vetted' is '2', then only trusted and unreviewed content will be returned (untrusted content will not be returned). If 'vetted' is '3', then only unreviewed content will be returned. If 'vetted' is '4', then only untrusted content will be returned.The default is to return all content.

        pages_params.update({'id':",".join(eol_ids), 'vetted':1, 'common_names': 'true'})
        r1 = sess.get(url,params=pages_params, timeout=10)
        pages_params.update({'id':",".join(eol_ids), 'vetted':3, 'common_names': 'false'})
        r2 = sess.get(url,params=pages_params, timeout=10)
    except requests.exceptions.Timeout:
        warn('socket timed out - URL {}'.format(url))
        return
    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError) as error:
        warn('Data not retrieved because {}\nURL: {}'.format(error, url))
        return

    returned_ids = [int(i.popitem()[0]) for i in r1.json()]
    d1 = [i.popitem()[1] for i in r1.json()]
    d2 = [i.popitem()[1] for i in r2.json()]
    
    conversion = dict(zip(returned_ids, input_ids))
    warn(conversion)
    final_data = {id:{'vn':"", 'DOid':None} for id in input_ids}

    for data in d1:
        VN_en=""
        try:
            for vn in data['vernacularNames']:
                if (vn['language'] and vn['language'] == args.language):
                  if VN_en=="" or ('eol_preferred' in vn and vn['eol_preferred'] == 'true'):
                      VN_en = vn['vernacularName'] #pick the first one
        except KeyError:
            #no vernacular names array
            pass
        final_data[conversion[data['identifier']]]['vn'] = VN_en
    
    trustedpics = {data['identifier']:[data['dataObjects'][0]['dataObjectVersionID'], data['dataObjects'][0]['dataRating']] for data in d1 if 'dataObjects' in data and len(data['dataObjects'])}
    unreviewedpics = {data['identifier']:[data['dataObjects'][0]['dataObjectVersionID'], data['dataObjects'][0]['dataRating']] for data in d2 if 'dataObjects' in data and len(data['dataObjects'])}
    
    for k in returned_ids:
        c = conversion[k]
        if k != c:
            warn("WARNING: ID passed in ({}) is not the same as ID returned ({})".format(c, k))
        if k not in trustedpics and k not in unreviewedpics:
            pass
        elif k in trustedpics and k not in unreviewedpics:
            final_data[c]['DOid'] = trustedpics[k][0]
        elif k not in trustedpics and k in unreviewedpics:
            final_data[c]['DOid'] = unreviewedpics[k][0]
        else:
            #must pick one with best score
            if trustedpics[k][1]<unreviewedpics[k][1]:
                final_data[c]['DOid'] = unreviewedpics[k][0]
            else:
                final_data[c]['DOid'] = trustedpics[k][0]
    return final_data
            

parser = argparse.ArgumentParser(description='Save best EoL image (data object) id for each EOL page IDs')
parser.add_argument('csv_infile', nargs='?', type=argparse.FileType('r'),  default=sys.stdin, help='A file containing EOL page ids')
parser.add_argument('csv_outfile', nargs='?', type=argparse.FileType('w'), default=sys.stdout, help='The new csv file with best data object ID and common name appended')
parser.add_argument('--csvfield', '-c',  type=int, default=0, help='Find the EOL page ID in this field of each row')
parser.add_argument('--language', '-lang', default="en", help='The language to use for vernacular names')
parser.add_argument('--verbose', '-v', action="store_true", help='verbose: output extra non-essential info')
args = parser.parse_args()

#make a single http session, which we can tweak
s = requests.Session()
retries = Retry(total=5,
                backoff_factor=1,
                status_forcelist=[ 500, 502, 503, 504 ])
s.mount('http://', HTTPAdapter(max_retries=retries))
            
            
with args.csv_infile as f:
    with args.csv_outfile as out:
        reader = csv.reader(f)
        writer = csv.writer(out, quoting=csv.QUOTE_MINIMAL)
        ids = set()
        output=[]
        
        for row in reader:
            output.append(row)
            try:
                ids.update([int(row[args.csvfield-1])]) 
            except ValueError:
                warn("'{}' cannot be converted to an integer in row {}".format(row[args.csvfield-1], row))
            except IndexError:
                #EOL page ID is missing from this row
                pass
            if len(ids)==25:
                vals = getPageInfo(ids, s)
                for o in output:
                    try:
                        lookup = int(o[args.csvfield-1])
                        if lookup in vals:
                            writer.writerow(o + [vals[lookup]['DOid'], vals[lookup]['vn']])
                        else:
                            writer.writerow(o)                    
                    except ValueError:
                        pass
                del output[:]
                ids.clear() 
        if len(ids):
            vals = getPageInfo(ids, s)
            for o in output:
                try:
                    lookup = int(o[args.csvfield-1])
                    if lookup in vals:
                        writer.writerow(o + [vals[lookup]['DOid'], vals[lookup]['vn']])
                    else:
                        writer.writerow(o)                    
                except ValueError:
                    pass
        

File: OZprivate/ServerScripts/Utilities/getBestEOLdoID.pl
#!/usr/bin/perl -ws
# Given a csv input file containing EOL page numbers, return a file of the best pd|cc-by|cc-by-sa picture ids and the common names
# we must then map the EOL ids against the OTT ids using ordered_leaves
# set -c=X to read EOL id from column X of the CSV file
#produce e.g. list of all leaves using 
# cut -f 2,4 -d, data/DBinputs/ordered_leaves.csv | ServerScripts/Utilities/getBestEOLdoID.pl -c=2 - > output.txt

# The simplest use of this script is e.g. 
#      ./getBestEOLdoID.pl pageIDs.txt output.txt

use strict;
use warnings;
use LWP::Simple;
use JSON -support_by_pp;
use Try::Tiny;
use Data::Dumper;
use Text::CSV;
use vars qw/$c/;  #column number for EOL ID
$c=1 unless defined $c;

## Default settings ##
my $APIkey="0e8786f5d94e9587e31ed0f7703c9a81f3036c7f"; #replace with your own API key.
my $vernacularLANG = 'en';
$vernacularLANG = $main::language if(defined($main::language));

my $pagesbase = "http://eol.org/api/pages/1.0.json"; #see http://eol.org/api/docs/pages
my %pages_params = (
  batch      => 'true',
#  key        => $APIkey,
  images     => 1,     # only look through the first 10 images
  videos     => 0,
  sounds     => 0,
  text       => 0,
  cache_ttl  => 2500, #only cache for 10 secs
  licenses   =>'pd|cc-by|cc-by-sa', #change this to get objects distributed under different licences
  iucn       => 'false',
  common_names=>'true',
  details    => 'false',
  taxonomy    => 'false',
  vetted     => 2, #If 'vetted' is given a value of '1', then only trusted content will be returned. If 'vetted' is '2', then only trusted and unreviewed content will be returned (untrusted content will not be returned). The default (0) is to return all content.
);
  

## Program starts here ##

die("The name of a file of EOL page ids names (or - for stdin) is required as a first argument\n") unless ($ARGV[0]);
my @taxa =(); #store all the data in here
my $pageids;
if ($ARGV[0] eq "-") {
  $pageids = \*STDIN;
} else {
  open($pageids, "<", $ARGV[0])
    or die "cannot open ".$ARGV[0]." for reading: $!";
}

if ($ARGV[1] && $ARGV[1] ne "-") {
  open(OUTPUT, ">:utf8", $ARGV[1])
    or die "cannot open ".$ARGV[1]." for (over)writing: $!";
} else {
  *OUTPUT = *STDOUT;
  binmode OUTPUT, ":utf8";
}
select OUTPUT;
$| = 1;


my @rows;
my $csv = Text::CSV->new ( { binary => 1 } )  # should set binary attribute.
                 or die "Cannot use CSV: ".Text::CSV->error_diag ();
 
while ( my $row = $csv->getline( $pageids ) ) {
     push @rows, $row;
}
$csv->eof or $csv->error_diag();
close $pageids;

my $h = {};

my @BATCHES;
push @BATCHES, [ splice @rows, 0, 10 ] while @rows;
foreach my $batch (@BATCHES) {
    print(join("\t", (@$_, getPageInfo(map {$_->[$c-1]} @$batch));
}


open(my $fh2, "<", $ARGV[1]);
while(<$fh2>) {
    s/^,(\d+)/,$h->{$1}/;
    print;
}


sub getPageInfo {
  my $ids = shift;  
  my $eol_data = {};
  my $url = $pagesbase."?id=".join("%2C",@$ids)."&".join("&", map{"$_=$pages_params{$_}"} keys %pages_params);
  my $i=0;
  my $pg;
  while (not($pg = fetch_json_page($url))) {
    last if (++$i==10); #try getting the page a few times
    sleep(2); #wait a bit and try again
  };
  unless ($pg) {
    warn "Error in getting json page result from EoL for url = '$url', tried $i times\n";    
  } else {
    #sort out names etc.
    if (ref($pg) ne "ARRAY") {
    	die "Error in EoL json page for url = '$url': no labelled data\n";    
    };
    foreach my $id (@$pg) {
        my @ids = keys %$id;
        die("No single id\n") if @ids != 1;
        my $EOLid = $ids[0];
        my $DOid = "";
        my $VN_en = "";
        if (!($id->{$EOLid}->{dataObjects}) || 0==@{$id->{$EOLid}->{dataObjects}}) {
          #warn "(no appropriate data objects found in EoL for $data{name} at ".pageID2URL($id)."\n";
        } else {
          $DOid = $id->{$EOLid}->{dataObjects}->[0]->{dataObjectVersionID}
        };

        foreach my $vn (@{$id->{$EOLid}->{vernacularNames}}) {
            if ($vn->{language} && $vn->{language} eq $vernacularLANG) {
              if (!$VN_en || ($vn->{eol_preferred} && $vn->{eol_preferred} eq 'true')) {
                $VN_en = $vn->{vernacularName}; #pick the first one
              }
            }
        }
        push(@$retvals, [$DOid, $VN_en]);
  };
 }
 return($retvals);
}


sub fetch_json_page
{
  my $json = new JSON;
  my ($json_url) = shift;
  # download the json page:
  my $json_text;
  my $content = get( $json_url );
  if (defined($content)) {
    try {
    # these are some nice json options to relax restrictions a bit:
      $json_text=$json->allow_nonref->utf8->relaxed->escape_slash->loose->allow_singlequote->allow_barekey->decode($content);
    } catch {
      warn "In string \"$content\" - JSON error: $_ \n";
    };
  };
  return $json_text;
}

sub trimname {
  #take an author name and make it nicer
  my $name = decode_entities(shift);
  $name =~ s/^\s*creator:?//i; #common idiom to start wikicommons authors with e.g. Creator: Joseph Smit
  $name =~ s/^\s*//;
  $name =~ s/\s*$//;
  $name =~ s/^(?:<.+?>)*unknown(?:<.+?>)*$//i; #strip "unknown" and tags
  return $name;
}